File: DP/2579.py
# https://www.acmicpc.net/problem/2579
# Name: 계단 오르기
# Topic: Dynamic Programming


def solution(stair):
    N = len(stair)
    dic = {1: stair[0], 2: stair[0]+stair[1]}
    for i in range(3, N + 1):
        if i == 3:
            dic[i] = max(stair[1]+stair[2], stair[0]+stair[2])
        else:
            dic[i] = max(dic[i - 3] + stair[i - 2] + stair[i - 1], dic[i - 2] + stair[i - 1])
    return dic[N]

if __name__ == '__main__':
    n = int(input())
    a = []
    for i in range(n):
        a.append(int(input()))

    print(solution(a))

File: DP/2579.cpp
// https://www.acmicpc.net/problem/2579
// Name: 계단 오르기
// Topic: Dynamic Programming

#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int main() {
    int num;
    cin >> num;
    vector<int> stair;
    for (int i = 0; i < num; i++) {
        int temp;
        cin >> temp;
        stair.push_back(temp);
    }

    vector<int> v(num);
    v[0] = stair[0], v[1] = stair[0] + stair[1], v[2] = max(stair[0]+stair[2], stair[1] + stair[2]);
    for (int i = 3; i < num; i++) {
        v[i] = max(v[i - 2] + stair[i], v[i - 3] + stair[i - 1] + stair[i]);
    }
    cout << v[num - 1];
}